base/src/MotionVectorExtractor.cpp

```diff
@@ -0,0 +1,279 @@
+				av_packet_unref(pkt);
+				av_frame_unref(avFrame);
+			}
+		}
+		return 0;
+	}
+public:
+	int mWidth = 0;
+	int mHeight = 0;
+	std::string rawFramePinId;
+private:
+	AVFrame* avFrame = NULL;
+	AVCodecContext* decoderContext = NULL;
+	std::function<frame_sp(size_t)> makeFrame;
+	std::function<frame_sp(size_t size, string& pinId)> makeFrameWithPinId;
+	bool sendDecodedFrame = false;
+};
+
+
+MotionVectorExtractor::MotionVectorExtractor(MotionVectorExtractorProps props) : Module(TRANSFORM, "MotionVectorExtractor", props)
+{
+	mDetail.reset(new Detail(props, [&](size_t size) -> frame_sp {return makeFrame(size); }, [&](size_t size, string& pinId) -> frame_sp { return makeFrame(size, pinId); }));
+	auto outputMetadata = framemetadata_sp(new FrameMetadata(FrameMetadata::MOTION_VECTOR_DATA));
+	rawOutputMetadata = framemetadata_sp(new RawImageMetadata());
+	motionVectorPinId = addOutputPin(outputMetadata);
+	mDetail->rawFramePinId = addOutputPin(rawOutputMetadata);
+}
+
+bool MotionVectorExtractor::init()
+{
+	mDetail->initDecoder();
+	return Module::init();
+}
+
+bool MotionVectorExtractor::term()
+{
+	return Module::term();
+}
+
+bool MotionVectorExtractor::validateInputPins()
+{
+	if (getNumberOfInputPins() != 1)
+	{
+		LOG_ERROR << "<" << getId() << ">::validateInputPins size is expected to be 1. Actual<" << getNumberOfInputPins() << ">";
+		return false;
+	}
+
+	framemetadata_sp metadata = getFirstInputMetadata();
+	FrameMetadata::FrameType frameType = metadata->getFrameType();
+	if (frameType != FrameMetadata::H264_DATA)
+	{
+		LOG_ERROR << "<" << getId() << ">::validateInputPins input frameType is expected to be H264_DATA. Actual<" << frameType << ">";
+		return false;
+	}
+	return true;
+}
+
+bool MotionVectorExtractor::validateOutputPins()
+{
+	auto size = getNumberOfOutputPins();
+	if (getNumberOfOutputPins() > 2)
+	{
+		LOG_ERROR << "<" << getId() << ">::validateOutputPins size is expected to be 2. Actual<" << getNumberOfOutputPins() << ">";
+		return false;
+	}
+
+	pair<string, framefactory_sp> me; // map element	
+	auto framefactoryByPin = getOutputFrameFactory();
+	BOOST_FOREACH(me, framefactoryByPin)
+	{
+		FrameMetadata::FrameType frameType = me.second->getFrameMetadata()->getFrameType();
+		if (frameType != FrameMetadata::MOTION_VECTOR_DATA && frameType != FrameMetadata::RAW_IMAGE)
+		{
+			LOG_ERROR << "<" << getId() << ">::validateOutputPins input frameType is expected to be MOTION_VECTOR_DATA or RAW_IMAGE. Actual<" << frameType << ">";
+			return false;
+		}
+	}
+
+	return true;
+}
+
+bool MotionVectorExtractor::shouldTriggerSOS()
+{
+	return mShouldTriggerSOS;
+}
+
+bool MotionVectorExtractor::process(frame_container& frames)
+{
+
+	auto inFrame = frames.begin()->second;
+	frame_sp motionVectorFrame;
+	frame_sp decodedFrame;
+	mDetail->getMotionVectors(inFrame, motionVectorFrame, decodedFrame);
+
+	frames.insert(make_pair(motionVectorPinId, motionVectorFrame));
+	frames.insert(make_pair(mDetail->rawFramePinId, decodedFrame));
```

Comment:
so this will always add an output pin even if "sendDecodedFrame" is false ?

Reply:
Improved it to send the decodedFrame only when sendDecodedFrame is true.